File: sync.py
# This allows gevent to be imported from the ./lib diretory
import os
dir_path = os.path.dirname(os.path.realpath(__file__))
import sys
sys.path.append(dir_path + "/lib/")

import gevent.monkey
gevent.monkey.patch_all()
import gevent

from hesburgh import heslog, hesutil
import json
import re
import shared
import urllib2
import time

# Issue: the first response that returns the api limit header will have actually gone through
# successfully - all the subsequent calls will fail until the limit is lifted. This is great
# except that we're in threads so we don't know if we're the first one to hit the error or not
# so we retry everything. If we retry a previously successful call we'll now get a version mismatch error

# The contentful calls per second is 10, so do 5 at a time (update + publish = 10 max calls) to get around ^ issue
concurentItems = 5


def updateItem(item):
  fields = item.get("fields", {})

  alephNumber = item.get("fields", {}).get("alephSystemNumber", {}).get("en-US")
  sysId = item.get('sys', {}).get('id')

  if not alephNumber:
    heslog.warn("No Aleph number for %s" % (sysId))
    return None

  alephItem = shared.getAleph(alephNumber)
  if not alephItem:
    heslog.warn("Couldn't find item %s" % alephNumber)
    return None

  if shared.isDifferent(alephItem, fields):
    version = item.get("sys", {}).get("version", 1)

    updated = shared.updateContentful(sysId, version, alephItem, fields)
    if not updated:
      return
    shared.publishContentful(sysId, updated.get("sys", {}).get("version", 1))


def run(event, context):
  heslog.addLambdaContext(event, context)

  start = 0
  total = 100

  headers = {
    "Authorization": "Bearer %s" % hesutil.getEnv("OAUTH", throw=True),
  }
  baseUrl = hesutil.getEnv("CONTENTFUL_QUERY_URL", throw=True)

  items = []
  while start < total:
    url = baseUrl + "&skip=%s" % (start)

    req = urllib2.Request(url, headers=headers)
    response = ""
    try:
      response = urllib2.urlopen(req)
    except urllib2.HTTPError as e:
      heslog.error("%s" % e.code)
      heslog.error(e.read())
      return None
    except urllib2.URLError as e:
      heslog.error(e.reason)
      return None

    responseObj = json.loads(response.read())

    start = responseObj.get("skip", 0) + responseObj.get("limit", 100)
    total = responseObj.get("total", 0)

    responseItems = responseObj.get("items", [])
    for item in responseItems:
      if item.get("sys", {}).get("archivedAt", None) is None:
        items.append(item)

  heslog.info("Updating %s items" % len(items))

  start = 0

  timer = hesutil.Timer()
  while len(items) > 0:
    threads = []
    use = items[:concurentItems]
    items = items[concurentItems:]

    heslog.info("Processing items %s through %s" % (start, start + len(use) - 1))
    start += concurentItems
    timer.start()
    for item in use:
      threads.append(gevent.spawn(updateItem, item))

    gevent.joinall(threads)
    dt = timer.end()
    # sleep to make it less likely we hit the contentful api limit
    time.sleep(1)


  heslog.info("Returning 200 OK")
  return { "statusCode": 200, "body": "" }

File: src/sync.js
const fetch = require('node-fetch')
const sleep = require('util').promisify(setTimeout)
const { t: typy } = require('typy')
const { getAleph, isDifferent, updateContentful, publishContentful } = require('./shared/helpers')
const { successResponse } = require('./shared/response')
const { sentryWrapper } = require('./shared/sentryWrapper')

// Issue: the first response that returns the api limit header will have actually gone through
// successfully - all the subsequent calls will fail until the limit is lifted. This is great
// except that we're in threads so we don't know if we're the first one to hit the error or not
// so we retry everything. If we retry a previously successful call we'll now get a version mismatch error

// The contentful calls per second is 10, so do 5 at a time (update + publish = 10 max calls) to get around ^ issue
const concurrentItems = 5

const updateItem = async (item) => {
  const alephNumber = typy(item, 'fields.alephSystemNumber.en-US').safeString
  const sysId = typy(item, 'sys.id').safeString

  if (!alephNumber) {
    console.warn(`No Aleph number for ${sysId}`)
    return
  }

  const alephItem = await getAleph(alephNumber)
  if (!alephItem) {
    console.warn(`Couldn't find item ${alephNumber}`)
    return
  }

  if (isDifferent(alephItem, item.fields)) {
    const version = typy(item, 'sys.version').safeNumber

    const updated = await updateContentful(sysId, version, alephItem, item.fields)
    if (!updated) {
      // Return the sysId so we know which records were processed, even though there are no changes
      return sysId
    }
    return publishContentful(sysId, typy(updated, 'sys.version').safeNumber)
  }
}

// Process items
const queueUpdates = async (items, start) => {
  const promises = []
  const use = items.slice(start, start + concurrentItems)

  console.log(`Processing items ${start} through ${start + use.length - 1}.`)

  use.forEach(item => {
    promises.push(updateItem(item))
  })

  return Promise.allSettled(promises).then(results => {
    const successful = []
    const errored = []

    results.forEach(result => {
      if (result.status === 'fulfilled') {
        if (result.value) {
          successful.push(result.value)
        }
      } else {
        const errorObj = result.reason.status
          ? {
            url: result.reason.url,
            status: result.reason.status,
            statusText: result.reason.statusText,
          }
          : result.reason
        errored.push(errorObj)
      }
    })
    return {
      successful: successful,
      errors: errored,
    }
  })
}

module.exports.handler = sentryWrapper(async (event, context, callback) => {
  // Build request
  const headers = {
    Authorization: `Bearer ${process.env.CONTENTFUL_MANAGEMENT_TOKEN}`,
  }
  const url = `${process.env.CONTENTFUL_CMA_URL}/entries?content_type=resource&order=sys.updatedAt&limit=1000`

  const response = await fetch(url, { headers: headers }).then(res => res.json())

  // Make list of items; exclude archived records
  const publishedItems = typy(response, 'items').safeArray.filter(item => {
    return typy(item, 'sys.archivedAt').isNullOrUndefined
  })

  console.log(`Updating ${publishedItems.length} items`)

  const promiseResults = {
    successful: [],
    errors: [],
  }

  let start = 0
  // wait before next call to make it less likely we hit the contentful api limit
  while (start < publishedItems.length) {
    const chunkResults = await queueUpdates(publishedItems, start)
    console.log('Chunk results:', JSON.stringify(chunkResults, null, 2))
    promiseResults.successful = promiseResults.successful.concat(chunkResults.successful)
    promiseResults.errors = promiseResults.errors.concat(chunkResults.errors)

    start += concurrentItems
    await sleep(1000) // Delay before next iteration to reduce risk of hitting API limits
  }

  return successResponse(callback, promiseResults, (promiseResults.errors.length > 0 ? 207 : 200))
})
